### ISSUE TITLE
Request - Client: Ignore JsonIgnore and readonly properties

### ISSUE THREAD

**Issue by weitzhandler** [TrackableEntities/trackable-entities#184]
Hi,

Let's have the CTC ignore any:

- Read only property
- Properties attributed as `JsonIgnore`

Can I PR?

**Comment by weitzhandler** [comment]
Fixes #165

**Comment by tonysneed** [comment]
JsonIgnore has to do with serialization.  Can you explain more about your scenario where you need this? If you like, you can create a PR that has just a failing test.

**Comment by weitzhandler** [comment]
I'm not sure I can come up with a failing test for it now, but the logic behind this is very simple:
If the client says `JsonIgnore` this property, meaning it doesn't want it to be serialized to the server, and in fact, this property may not even exist in the server (for example that property is only declared in a client-only partial class), thus, it shouldn't be in the `ModifiedProperties` either.

**Comment by weitzhandler** [comment]
Imaging [this]() class would look like this in the client:

using System;
using System.Collections.Generic;
using Newtonsoft.Json;

```
namespace TrackableEntities.Client.Tests.Entities.FamilyModels
{
    [JsonObject]
    public partial class Child : EntityBase
    {
        public Child() { }
        public Child(string name)
        {
            _name = name;
        }

        private string _name;
        public string Name
        {
            get { return _name; }
            set
            {
                if (value == _name) return;
                _name = value;
                NotifyPropertyChanged(() => Name);
            }
        }

        [JsonIgnore]
        public string DisplayName
        {
            get
            {
                return Name;
            }
            set 
            {
                Name = value;
            }
        }

        public string ShortName
        {
            get
            {
                return Name == null ? null : Name.Substring(0, Math.Max(Name.Length, 3));
            }
        }

        private ChangeTrackingCollection<Child> _children;
        public ChangeTrackingCollection<Child> Children
        {
            get { return _children; }
            set
            {
                if (Equals(value, _children)) return;
                _children = value;
                NotifyPropertyChanged(() => Children);
            }
        }
    }
}
```

Now, the readonly property (`ShortName`) does need a property changed notification so to affect UI, but TE isn't supposed to care about it.
The `DisplayName` property, is a client-only property, and it's not even in the server. Why should it exist in the `ModifiedProperties` property. Now this has already been fixed with the non-mapped members issue, but still, the server isn't supposed to hear about these properties, if they were chosen to be readonly (meaning non-modifiable) or ignored from the serialization by `JsonIgnore`ing them.

**Comment by tonysneed** [comment]
Have you tried ExcludedProperties on CTC?

**Comment by weitzhandler** [comment]
Never heard of it... Lemme check.
But anyway, why don't we make this automatic?

**Comment by tonysneed** [comment]
Yes, so I would suggest an extension method that accepts a lambda (strategy pattern).  It can be a Func<T, bool>.  Then the caller can apply whatever strategy they want.  If the delegate returns true, then the method can add the property to `ExcludedProperties`.  The method can be called ExcludeProperties.  The T can be an expression where you can specify a property.

The caller can then apply whatever logic they want (read-only, JsonIgnore, no DataMember attribute, etc.

**Comment by weitzhandler** [comment]
Makes sense.
Should we make a default method that ignores readonly and `JsonIgnore` props?

**Comment by weitzhandler** [comment]
Makes sense.
Should we make a default method that ignores readonly and `JsonIgnore` props, so that we don't have to call the default ignores for each CTC manually?

**Comment by tonysneed** [comment]
Let's start with the generic method and go from there, perhaps adding overloads for common patterns.

**Comment by weitzhandler** [comment]
@tonysneed 
>The method can be called ExcludeProperties. The T can be an expression where you can specify a property.

We need a way that covers more than one property.
I'd say it should be `Func<TEntity, IEnumerable<string>> InitExcludedProperties { get; set; }`, that when is not null, fills the `ExcludedProperties` with the provided list, if we're going about adding that property.
Besides, the excluded properties in the CTC is for the root entity type. How will we access the entire graph's `ExcludedProperties` dynamically?

That's why I'm personally still voting for ignoring `readonly` and `JsonIgnore`d properties as a means of dynamically excluding properties, rather than rely on manually messing with the `ExcludedProperties` property of the CTC.

**Comment by tonysneed** [comment]
> We need a way that covers more than one property.

So the way `ExcludeProperties` should work is that it iterates over all the properties of the CTC entity or entities.  We can start with strings for property names.  (We can also implement recursion if `recursive` parameter is true, by walking down the object graph.)

```csharp
private void ExcludeProperties(TEntity entity, Func<string, bool> predicate, bool recursive)
{
    foreach (var prop in typeof(TEntity).GetRuntimeProperties())
    {
        if (predicate(prop.Name))
        {
            ExcludedProperties.Add(prop.Name);
        }
    }
}
```

Here is the public method that calls the private method.

```csharp
public void ExcludeProperties(Func<string, bool> predicate, bool recursive)
{
    foreach (var item in Items)
    {
        ExcludeProperties(item, predicate, recursive);
    }
}
```

How does this look?

Cheers,
Tony

**Comment by weitzhandler** [comment]
The problem with setting up the excluded properties in the CTC is that it must be performed on each CTC individually throughout the entire graph, which doesn't really make sense.
There has to be a global way to configure excluded properties per entity type.

**Comment by tonysneed** [comment]
The recursive flag is there to allow you to do this. We just need to add the CTC type as an argument on the Func, so that the callback knows the CTC type. The approach here is to build out the low level API based on a Strategy pattern, then build a higher level of abstraction on top of that, so you can do things like look for Json attributes.

**Comment by weitzhandler** [comment]
I'll try to implement during the week.
I really appreciate you kind and detailed responses, and your sympathy!

**Comment by weitzhandler** [comment]
> It's just that I'd like to try for a more extensible design

The config can't sit in the CTC. It has to be configured in a global map, via attributes or other per-class/property excludes.
